core/src/main/java/org/apache/iceberg/io/ImmutableStorageCredential.java

```diff
@@ -0,0 +1,359 @@
+      return this;
+    }
+
+    /**
+     * Put one entry to the {@link StorageCredential#config() config} map. Nulls are not permitted
+     *
+     * @param entry The key and value entry
+     * @return {@code this} builder for use in a chained invocation
+     */
+    @CanIgnoreReturnValue
+    public final Builder putConfig(Map.Entry<String, ? extends String> entry) {
+      String k = entry.getKey();
+      String v = entry.getValue();
+      this.config.put(
+          Objects.requireNonNull(k, "config key"),
+          Objects.requireNonNull(v, v == null ? "config value for key: " + k : null));
+      return this;
+    }
+
+    /**
+     * Sets or replaces all mappings from the specified map as entries for the {@link
+     * StorageCredential#config() config} map. Nulls are not permitted
+     *
+     * @param entries The entries that will be added to the config map
+     * @return {@code this} builder for use in a chained invocation
+     */
+    @CanIgnoreReturnValue
+    public final Builder config(Map<String, ? extends String> entries) {
+      this.config.clear();
+      return putAllConfig(entries);
+    }
+
+    /**
+     * Put all mappings from the specified map as entries to {@link StorageCredential#config()
+     * config} map. Nulls are not permitted
+     *
+     * @param entries The entries that will be added to the config map
+     * @return {@code this} builder for use in a chained invocation
+     */
+    @CanIgnoreReturnValue
+    public final Builder putAllConfig(Map<String, ? extends String> entries) {
+      for (Map.Entry<String, ? extends String> e : entries.entrySet()) {
+        String k = e.getKey();
+        String v = e.getValue();
+        this.config.put(
+            Objects.requireNonNull(k, "config key"),
+            Objects.requireNonNull(v, v == null ? "config value for key: " + k : null));
+      }
+      return this;
+    }
+
+    /**
+     * Builds a new {@link ImmutableStorageCredential ImmutableStorageCredential}.
+     *
+     * @return An immutable instance of StorageCredential
+     * @throws java.lang.IllegalStateException if any required attributes are missing
+     */
+    public ImmutableStorageCredential build() {
+      if (initBits != 0) {
+        throw new IllegalStateException(formatRequiredAttributesMessage());
+      }
+      return ImmutableStorageCredential.validate(
+          new ImmutableStorageCredential(prefix, createSerializableMap(false, false, config)));
+    }
+
+    private String formatRequiredAttributesMessage() {
+      List<String> attributes = new ArrayList<>();
+      if ((initBits & INIT_BIT_PREFIX) != 0) attributes.add("prefix");
+      return "Cannot build StorageCredential, some of required attributes are not set "
+          + attributes;
+    }
+  }
+
+  private static <K, V> Map<K, V> createSerializableMap(
+      boolean checkNulls, boolean skipNulls, Map<? extends K, ? extends V> map) {
+    switch (map.size()) {
```

Comment:
This feels overly complicated for this use case.  Why not just return `emptyMap` or `MashMap`?  I don't think we need to carry the optimizations for a single entry map as that's incredibly unlikely and overly complicated.

Reply:
I've removed the optimizations